[WIP] Touch support

chayleaf opened the pull request:
See #122

This implements touch support in the following way:

- Holding the screen and moving your finger up/down scrolls down/up. Scroll speed is hardcoded, and the mouse position is set to the initial tap position.
- A short tap emulates a return button press.
- A long tap that didn't move far enough to scroll emulates an escape button press (ideally, `-click-to-exit` would work, then this would be unnecessary)
This is obviously not ideal, so I'd like some feedback on how a proper implementation should look like

Maybe programmable gesture support should be added? this would be useful for e.g. mode switching. I have no idea how the config for that should look though.

chayleaf committed:
wip touch support

chayleaf committed:
wip touch support

chayleaf committed:
still scroll after the tap moves outside the rofi window

chayleaf committed:
still scroll after the tap moves outside the rofi window

chayleaf committed:
exit on long tap

chayleaf committed:
exit on long tap

lbonn commented on source/wayland/display.c:
```diff
@@ -526,6 +526,100 @@ static const struct wl_keyboard_listener wayland_keyboard_listener = {
     .repeat_info = wayland_keyboard_repeat_info,
 };
 
+static void wayland_touch_down(void *data, struct wl_touch *wl_touch,
```
The indentation is a bit inconsistent with the rest of the file.

Could you please run clang-format on the lines you've added?

lbonn commented on source/wayland/display.c:
```diff
@@ -526,6 +526,100 @@ static const struct wl_keyboard_listener wayland_keyboard_listener = {
+    return;
+  }
+  int key = KEY_ENTER;
+  if (time - self->touches[id].start_time > 200) {
```
Maybe define this 200 as a constant with a proper name at the top of this function?

lbonn commented on source/wayland/display.c:
```diff
@@ -526,6 +526,100 @@ static const struct wl_keyboard_listener wayland_keyboard_listener = {
+}
+
+static int32_t y_offset_to_line_offset(int32_t y_offset) {
+  static const int32_t line_height = 20;
```
So, this "20" is some kind of hard-coded sensitivity?

It is fine for now but maybe add a TODO comment like "TODO: make sensitivity parametrizable"

lbonn reviewed (commented):
I cannot really test this right now but it seems fine to have it merged as a first iteration :)

Thanks!

in0ni commented:
Moving discussion from #156 here, and will close according to authors responses.

@chayleaf, do you plan continuing work on this? I figured your PR was exactly that, a request to pull this into code base, and @lbonn has already expressed he would merge as an initial iteration.

Though it was not clear from this thread that this does not work as you state in 156... yes it's not perfect, but from my initial testing on two devices suffices to avoid having to replace rofi. There are some minor bugs, but at least it's functional and for me it greatly helps. Better this than having to look for another launcher.

If you can provide some feedback, I'm happy to help, or close the other PR -- just let me know :)

PS: duly noted regarding history/authorship, will do, thanks.

chayleaf commented:
> do you plan continuing work on this?

I am daily driving this branch on my phone, which is why I know it's not ready to merge yet. I do want to fix the issues when I have the time.

> I figured your PR was exactly that, a request to pull this into code base

There is a reason this is a *draft* PR, draft PRs aren't ready for merging

For now, I recommend you to apply these patches locally.

chayleaf commented on source/wayland/display.c:
```diff
@@ -526,6 +526,100 @@ static const struct wl_keyboard_listener wayland_keyboard_listener = {
+  if (time - self->touches[id].start_time > 200) {
+    key = KEY_ESC;
+  }
+  nk_bindings_seat_handle_key(wayland->bindings_seat, NULL,
```
on some rofi launches, this doesn't work (i.e. if it doesn't work, it won't magically fix itself if you click multiple times). TODO figure out why

DaveDavenport committed:
Mark as dev version.

DaveDavenport committed:
Add rofi-actions to the configure.ac dist list

DaveDavenport committed:
Build manpage

DaveDavenport committed:
[Build] More autoconf fixes for dist.

DaveDavenport committed:
Release 1.7.9.1

DaveDavenport committed:
Mark as dev release again

lbonn committed:
Print some info about the display backend in --help

lbonn committed:
Set version to next's dev

lbonn committed:
Sync back a lot of the docs with main rofi

lbonn committed:
Fix one outdated copyright notice

DaveDavenport committed:
[Build] Remove autoconf tools.

DaveDavenport committed:
Add `-info` option.

Include wayland version.

isssue: #2133

DaveDavenport committed:
Remove GCOV from help/info output.

issue: #2133

lbonn committed:
Remove some remaining m4 utils

lbonn committed:
Remove libgwater autotools directives

lbonn committed:
Remove autotools CI steps

lbonn committed:
Remove travis configuration

lbonn committed:
Restore issue threads CI workflows

DaveDavenport committed:
Fix building error true->TRUE

lbonn committed:
Merge remote-tracking branch 'upstream/next' into wayland-next

lbonn committed:
Remove dead screenshot code

lbonn committed:
Clean up gitignore

lbonn committed:
Add xcb compile status in info

Also try to make the option printing more consistent

DaveDavenport committed:
[Issues] Add rofi -info to templates and documentation.

issue: #2134

DaveDavenport committed:
[Build] Update build instructions for enabling/disabling wayland/x11

Also by default they both need to be available, you can disable
depending on your needs.

DaveDavenport committed:
Remove ternary operator from function call argument.

lbonn committed:
Restore doxygen check on CI

lbonn committed:
Fix github actions (hopefully)

DaveDavenport committed:
[Build] Use auto-option for enabling/disabling wayland/xcb

lbonn committed:
Check all dependencies for xcb/wayland (#2142)

DaveDavenport committed:
[Doc] Update build instructions with changes to backend selection.

remove ref. to fork from README.

DaveDavenport committed:
[Wayland] If there is one monitor use dpi of that monitor.

lbonn committed:
[Wayland] HiDPI: Try the first transparent frame trick

lbonn committed:
Revert dpi detection hacks

- "[Wayland] HiDPI: Try the first transparent frame trick"
commit 146c7b16a09162fde688ee04a0b908f5fe139ce6.

- "[Wayland] If there is one monitor use dpi of that monitor."
commit 9afe5434ea5b8b66fbe540ba676540a3a784e179.

lbonn committed:
[Wayland] Estimate DPI from one detected monitor (#2148)

* [Wayland] Estimate DPI from one detected monitor

This will work for single-monitor setup

Co-Authored-by: lbonn

* [Wayland] Update the DPI estimation to also use monitor name if available.

* [Doc] Update DPI setting for wayland in manpage.

---------

Co-authored-by: Qball <qball@blame.services>

lbonn committed:
Delay display failure exit after help prints

DaveDavenport committed:
Improve error message a bit

leavitals committed:
Use portable input-codes.h for mouse button constants

Adds a new header to define BTN_LEFT/RIGHT/MIDDLE for cross-platform
support. Simplifies display.c by removing Linux-specific checks.

DaveDavenport committed:
[Doc] Improve documentation for -theme.

Issue: #2152

DaveDavenport committed:
[Rofi]  Move setup_abe before config parsing, so config options are added to system.

Issue: #2163

DaveDavenport committed:
start 2.0.0 release notes

DaveDavenport committed:
releasenotes 2.0.0 -  small updates about known issues.

DaveDavenport committed:
[DOC] Remove make from INSTALL.md

DaveDavenport committed:
Update releasenotes.

DaveDavenport committed:
Update releasenotes

DaveDavenport committed:
Small tweak.

DaveDavenport committed:
[DOC] Add noanim rule for hyprland to manpage

DaveDavenport committed:
[Keyb] Fix typo

Issue: #2170

DaveDavenport committed:
[GitHub] Add some clarification to template

DaveDavenport committed:
Merge branch 'releasenotes-2.0.0' into next

DaveDavenport committed:
Release 2.0.0-beta1

DaveDavenport committed:
2.0.0 release.

chayleaf committed:
mobile: fix build
